### DISCUSSION TITLE
ApolloServer instance doesn't have appropriate @cacheControl set

### DISCUSSION THREAD

**Discussion by itsnotrisky** [apollographql/apollo-server#6747] [Q&A]
I have my ApolloServer instance setup as follow:

```js
const resolvers = mergeResolvers(resolversArray)
const typeDefs = mergeTypeDefs(typesArray)
const apollo = new ApolloServer({
    cache: new BaseRedisCache({
      client: redis,
    }),
    resolvers,
    typeDefs,
})
```
The problem is any `@cacheControl` directives don't show up on SDL Schema from sandbox where `@deprecated` can be seen - thus none of query cached because the hint is not there. Also I can see `typeDefs` does output appropriate cacheControl hint to be fed to ApolloServer. How do I troubleshoot this?

**Comment by glasser** [comment]
GraphQL introspection does not have a mechanism for showing the
applications of schema directives.

**Reply by itsnotrisky** [reply]
Sorry I was referring to the SDL menu from Schema on the sandbox page https://imgur.com/a/D65Av72
I can see the `@cacheControl` declaration there but not on the fields/types I have set, even though there are `@deprecated` directives shown there.

**Reply by glasser** [reply]
Right, but Sandbox only knows what your server's schema is by querying it over the introspection API, and the GraphQL introspection API doesn't have a general mechanism for showing the applications of schema directives.

**Reply by itsnotrisky** [reply]
I'm trying to emphasize that the ApolloServer instance doesn't "output" cacheControl directives even though it's been properly fed into typeDefs and I thought by saying introspection from sandbox was the right place since it shows other directives as well in the SDL schema.

I think that's where my issue lies - nothing gets cached because the directive somehow stripped off unless I set `defaultMaxAge`.

Hopefully that make sense, I'll change the title so that it's more about the directive issue than introspection.

**Reply by glasser** [reply]
Yes, introspection allows Sandbox to learn that directives are defined in your schema, but not if they are applied to schema elements. It's a missing piece of GraphQL introspection itself.

**Comment by itsnotrisky** [comment]
I had a try to set defaultMaxAge to a value and it did work providing `cache-control` header for all requests ignoring the individual hints I have set, it didn't appear on the SDL as well.
```
ApolloServerPluginCacheControl({
  defaultMaxAge: 10,
}),
```
Here's the schema I have from code then translated to `typeDefs` for ApolloServer and how it looks on the sandbox
<details>
<summary>Traction.graphql</summary>

```
type CampaignTractionPayload @cacheControl(maxAge: 300) {
  traction: CampaignTraction
}
```
</details> 
<details>
<summary>typeDefs output</summary>

```
{
  "kind": "ObjectTypeDefinition",
  "name": {
    "kind": "Name",
    "value": "CampaignTractionPayload",
    "loc": {
      "start": 188,
      "end": 211
    }
  },
  "interfaces": [],
  "directives": [
    {
      "kind": "Directive",
      "name": {
        "kind": "Name",
        "value": "cacheControl",
        "loc": {
          "start": 213,
          "end": 225
        }
      },
      "arguments": [
        {
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "maxAge",
            "loc": {
              "start": 226,
              "end": 232
            }
          },
          "value": {
            "kind": "IntValue",
            "value": "300",
            "loc": {
              "start": 234,
              "end": 237
            }
          },
          "loc": {
            "start": 226,
            "end": 237
          }
        }
      ]
    }
  ]
}
```
</details> 

<details>
<summary>SDL from sandbox</summary>

```
type CampaignTractionPayload {
  traction: CampaignTraction
}
```
</details>
